lib/invoker.js

```diff
@@ -24,21 +25,53 @@ const invoke = (userFunction) => {
     };
 };
 
-module.exports = (userFunctionUri, port) => {
-    const userFunction = lookUpFunction(userFunctionUri);
-
-    const server = new grpc.Server();
-    server.addService(services.RiffService, {invoke: invoke(userFunction)});
+const executeInit = async (initHook, timeout, server) => {
+    logger('Calling $init hook (if any)');
+    return await guardWithTimeout(initHook, timeout)
+        .then(() => logger('$init hook successfully completed (or skipped if none)'))
+        .catch((err) => {
+            logger(`Error during $init hook execution: ${err.toString()}`);
+            logger('Shutting down now');
+            server.forceShutdown();
+            return Promise.reject('$init error');
+        });
+};
 
+const registerShutdown = (globals, destroyHook, timeout, server) => {
     const shutdown = () => {
         logger('Graceful shutdown started');
-        server.tryShutdown(() => logger('Graceful shutdown completed'));
+        logger('Calling $destroy hook (if any)');
+
+        guardWithTimeout(destroyHook, timeout)
+            .then(() => {
+                logger('$destroy hook successfully completed (or skipped if none)');
+                server.tryShutdown(() => logger('Graceful shutdown completed'));
+            })
+            .catch((err) => {
+                logger(`Error during $destroy hook execution: ${err.toString()}`);
+                logger('Shutting down now');
+                server.forceShutdown();
+            });
     };
-    process.on('SIGTERM', shutdown);
-    process.on('SIGINT', shutdown);
+    globals.process.once('SIGTERM', shutdown);
+    globals.process.once('SIGINT', shutdown);
+};
+
+module.exports = async (globals, userFunctionUri, options) => {
```

Comment:
`globals` is too close to `global` for comfort. Can we pick a different name? It's also not clear to be how globals are different from options.

Reply:
It's currently just a way to not rely on the globally available `process` directly (which would make testing much harder). `options` specify invoker configuration options.
Happy to rename it to `globalScope` or something like that.

Reply:
it's the "global" aspect that I object most to. If it's desirable to decouple from the process object to facilitate testing, then I'd suggest not using the process object (or even it's facade).